sky/skylet/log_lib.py

```diff
@@ -448,7 +463,26 @@ def tail_logs(job_id: Optional[int],
         try:
             start_stream = False
             with open(log_path, 'r', encoding='utf-8') as f:
-                for line in f.readlines():
+                if tail > 0:
+                    # If tail > 0, we need to read the last n lines.
+                    # We use double ended queue to rotate the last n lines.
+                    lines = collections.deque(f.readlines(), maxlen=tail)
+                    # Check the first few lines to decide if streaming should
+                    # start. If start_stream_at is within these lines, delay
+                    # streaming.
+                    # Example: tail=115, file lines=120, print lines 9 to 120.
+                    # We can't simply print the entire lines object, as
+                    # start_stream_at is still within the lines.
+                    start_stream = True
+                    for index, line in enumerate(lines):
+                        if start_stream_at in line:
+                            start_stream = False
+                            break
+                        if index >= PEEK_HEAD_LINES_FOR_START_STREAM:
+                            break
```

Comment:
Can we double check if this logic is correct? It seems to me that it does not actually check the first lines of the file, but only the last lines?

That `start_stream_at` line should appear in the first lines of the file not the `lines` that are deque'ed.

Also, we should start streaming when `start_stream_at` is in the line, instead of setting it to False?

Reply:
This reminds me, we should add some smoke test for this functionality.

Reply:
The logic is to check the first few lines of the deque to see if it still contains the `start_stream_at` line and set `start_stream` accordingly.

Let's say:

* The file contains 30 lines, 1–30.
* lines contains 20 lines, 11–30.

If `start_stream_at` is line 10, it's not in deque `lines` because it's already dequeued. We need to print all the lines, so `start_stream` is True in this case. 

If `start_stream_at` is line 11, it's still in deque `lines`, so I set `start_stream` to False, and the original print code below will iterate through the lines, start printing only at line 11, skipping line 10.

I expect the line number of `start_stream_at` might change in the future, so I set `PEEK_HEAD_LINES_FOR_START_STREAM `to a constant. This means peeking at the head of the deque to see if `start_stream_at` is still in it.

I didn't construct a new list from the current deque to remove the lines before `start_stream_at` because it saves memory when files are large.

Reply:
The first case makes sense to me, but for the second case, it is possible that the `start_stream_at` have not shown up in the log file, e.g. the job has not print out the `start_stream_at` yet and will print it later. With the current second case handling, we will print all the output.

Reply:
Thanks @Michaelvll Good catch, Good catch—there exists a third case. In this case, `start_stream_at` hasn't been printed out yet. 

Now I will change the logic to:

* If `start_stream_at` is in the deque (case 1), set `start_stream` to `False`.
* If `start_stream_at` is not in the deque, peek at the head of the current log file:
    * If `start_stream_at` is already at the head of the current log file, set `start_stream` to `True` (case 2).
    * If `start_stream_at` is not at the head of the current log file, set `start_stream` to `False` (case 3).

case 3:

<img width="967" alt="image" src="https://github.com/user-attachments/assets/38121565-f6b6-4d96-8ff7-758cc6178e5b">